### PR TITLE
Mvp-287-createEndpoint-factory-function

### PR DESCRIPTION
**Pull request by SDToews**
## 🎯 Summary

  
[MVP-287](https://citz-imb.atlassian.net/jira/browse/MVP-287)

## 🔰 Checklist

- [x] I have read and agree with the following checklist and am following the guidelines in our [Code of Conduct](/CODE_OF_CONDUCT.md) document.

> - I have performed a self-review of my code.
> - I have commented my code, particularly in hard-to-understand areas.
> - I have made corresponding changes to the documentation where required.
> - I have tested my changes to the best of my ability.
> - I have consulted with the team if introducing a new dependency.
> - My changes generate no new warnings.

### CODE REVIEW

File: src/controller/controller.class.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      if (Array.isArray(allItems.data) && allItems.data.length === 0)
+        return res.status(httpStatusCode.NO_CONTENT).json(allItems);
+
+      res.status(httpStatusCode.OK).json(allItems);
+    });
+
+    this.getItemById = errorWrapper(async (req: Request, res: Response) => {
+      const item = await service.getItemById(req.params.id);
+
+      if (!item) return res.status(httpStatusCode.BAD_REQUEST).json({ message: 'Item not found' });
```
Comment (BradyMitch) [inline src/controller/controller.class.ts]:
```suggestion
      if (!item) return res.status(httpStatusCode.NOT_FOUND).json({ message: 'Item not found' });
```

File: src/controller/controller.class.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+      res.status(httpStatusCode.OK).json(item);
+    });
+
+    this.getItemByWhere = errorWrapper(async (req: Request, res: Response) => {
+      const item = await service.getItemByWhere(req.body);
+
+      if (!item) return res.status(httpStatusCode.BAD_REQUEST).json({ message: 'Item not found' });
```
Comment (BradyMitch) [inline src/controller/controller.class.ts]:
```suggestion
      if (!item) return res.status(httpStatusCode.NOT_FOUND).json({ message: 'Item not found' });
```

File: src/controller/controller.class.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      res.status(httpStatusCode.CREATED).json(createdItem);
+    });
+
+    this.createItems = errorWrapper(async (req: Request, res: Response) => {
+      const createdItems = await service.createItems(req.body);
+
+      res.status(httpStatusCode.CREATED).json(createdItems);
+    });
+
+    this.updateItem = errorWrapper(async (req: Request, res: Response) => {
```
Comment (BradyMitch) [inline src/controller/controller.class.ts:58]:
```suggestion
    this.updateItem = errorWrapper(async (req: Request, res: Response) => {
```

Comment (BradyMitch) [inline src/controller/controller.class.ts]: Can we rename these to updateItemById and deleteItemById ? Maybe even add a update and delete item by where

File: src/controller/controller.class.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      const updatedItem = await service.updateItem(req.params.id, req.body);
+
+      if (!updatedItem)
+        return res.status(httpStatusCode.BAD_REQUEST).json({ message: 'Item not found' });
```
Comment (BradyMitch) [inline src/controller/controller.class.ts]:
```suggestion
        return res.status(httpStatusCode.NOT_FOUND).json({ message: 'Item not found' });
```

File: src/endPoint/createEndpoint.factory.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// import { Controller } from '@/controller';
+// import { Service } from '@/service';
```
Comment (BradyMitch) [inline src/endPoint/createEndpoint.factory.ts]: Can you remove unused import statements

File: src/endPoint/endpoint.class.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { createController } from '@/controller';
+import { createRouter } from '@/router';
+// import { Service } from '@/service';
```
Comment (BradyMitch) [inline src/endPoint/endpoint.class.ts]:
```suggestion
```

File: src/router/index.ts [hunk 7]
```diff
@@ -0,0 +1,2 @@
+// export * from './router.class';
```
Comment (BradyMitch) [inline src/router/index.ts]:
```suggestion
```

File: src/router/router.class.ts
Comment (BradyMitch) [inline src/router/router.class.ts:1]: Is this file used?

File: src/router/router.factory.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import express from 'express';
+import { Controller } from '@/controller';
+
+export const createRouter = <TEntity>(dataController: Controller<TEntity>) => {
+  const router = express.Router();
+
+  router.get('/', dataController.getAllItems);
+  router.get('/:id', dataController.getItemById);
+  router.post('/', dataController.createItem);
+  router.put('/:id', dataController.updateItem);
```
Comment (BradyMitch) [inline src/router/router.factory.ts]: Is it updating an existing item or replacing the data of the existing item?
patch is used for updating and put is used when replacing the item

File: src/utils/HttpError.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export class HttpError extends Error {
```
Comment (BradyMitch) [inline src/utils/HttpError.ts:1]: We can import this from express-utilities once published but for now this is good.

File: src/router/router.factory.ts (hunk 8 above)
Comment (SDToews) [reply src/router/router.factory.ts]: This section will be refactored in the next ticket as it doesn't accurately reflect the path.  I'll keep your comment in mind.

File: src/utils/HttpError.ts (hunk 9 above)
Comment (SDToews) [reply src/utils/HttpError.ts:1]: indeed, I await patiently :)

File: src/router/router.class.ts
Comment (SDToews) [reply src/router/router.class.ts:1]: not yet, or maybe never, stay tuned for the next ticket.

File: src/endPoint/createEndpoint.factory.ts (hunk 5 above)
Comment (SDToews) [reply src/endPoint/createEndpoint.factory.ts]: can do

File: src/repository/repository.class.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+  getAllItems: () => Promise<EntitySchema<TEntity>[]>;
+
+  getItemById: (_id: string) => Promise<EntitySchema<TEntity> | null>;
+
+  getItemByWhere: (_where: object) => Promise<EntitySchema<TEntity> | null>;
+
+  createItem: (_item: EntitySchema<TEntity>) => Promise<EntitySchema<TEntity>>;
+
+  updateItem: (
```
Comment (BradyMitch) [inline src/repository/repository.class.ts]: Still have to update these to updateItemById and so on

File: src/controller/controller.class.ts (hunk 3 above)
Comment (SDToews) [reply src/controller/controller.class.ts]: I haven't finalized the default functionality yet, but I also want users to be able to add their own endpoint by configuration as well.

changed the names of update and delete by Id

File: src/repository/repository.class.ts (hunk 10 above)
Comment (SDToews) [reply src/repository/repository.class.ts]: done.

**Review by BradyMitch — APPROVED** [review]
Great work!

**Pull request merged**